### PR TITLE
[codex] feat(funasr-transcribe): add optimized ONNX transcription

### LINKED ISSUE #14: feat: funasr-transcribe 增加 SenseVoice-Small / ONNX 加速模式，支持单人讲课快速转录
## 背景

当前 funasr-transcribe 使用 Paraformer-large + 说话人分离（diarization），3小时音频在 Mac CPU 上转录耗时较长。虽然 RTF 低（~0.03），但说话人分离带来额外开销。

对于单人讲课/语音场景，不需要说话人分离，应该支持切换到更快的模式。

---

## 技术约束（重要）

**说话人分离（CAM++）目前只和 Paraformer 系列兼容**，原因：
- FunASR 说话人分离管线依赖时间戳（timestamp）信息
- Paraformer 输出精确的字级时间戳 → CAM++ 可据此标注每段话是谁说的
- SenseVoice 不输出时间戳 → CAM++ 拿不到必要信息 → 报错 `can not get timestamp`

**已知 Issue**：
- FunASR Issue #2706: SenseVoice + CAM++ compatibility
- FunASR Issue #2662: SenseVoice + CAM++ error
- FunASR Issue #2121: Timestamp requirement for diarization

截至目前，SenseVoice + 说话人分离尚未修复。

---

## 方案对比

| 场景 | 推荐方案 | 速度 |
|------|---------|------|
| 需要说话人分离（多人对话） | Paraformer-large（当前方案） | 基准速度 |
| 需要说话人分离 + 更快 | Paraformer-large ONNX + CAM++ | 快 2-4x |
| 单人讲课，不需要分离 | SenseVoice-Small ONNX | 最快（比 Paraformer 快 5x+） |
| 单人讲课，不需要分离 | Paraformer-large 关闭 diarize | 也快很多 |

---

### 1. 新增 SenseVoice-Small 模型支持

**模型特点**：
- FunASR 团队新模型，延迟 < 80ms
- 比 Whisper-Small 快 5 倍，比 Whisper-Large 快 15 倍
- 支持中文/英文/粤语/日语/韩语
- 额外支持情感识别、声学事件检测
- ⚠️ 不支持说话人分离（技术约束，见上方说明）

**适用场景**：单人讲课、单人语音笔记等已知是单人的场景

### 2. 新增 ONNX Runtime 加速

**实现方式**：
- 官方提供 `funasr-onnx` 包，支持 Paraformer 和 SenseVoice 的 ONNX 版本
- 支持 INT8 量化，进一步加速
- 安装：`pip install funasr-onnx`

### 3. 单人模式切换

用户在使用 skill 时，如果已知是单人讲课/语音，可以通过参数指定走快速路径：

```
# 单人快速模式（自动使用 SenseVoice，不启用 diarization）
转录这个音频，单人讲课

# 普通模式（保持原有 Paraformer + diarization）
转录这个音频，有多人讨论
```

### 4. 技术方案

**服务端接口变更**：
```python
POST /transcribe
{
  "file_path": "xxx.m4a",
  "model": "sensevoice",    # 可选：sensevoice | sensevoice-onnx | paraformer（默认）
  "diarize": false          # 单人模式时自动设为 false
}
```

**新增 ONNX 服务模式**（新增 `scripts/server-onnx.py`）：
- 启动时加载 ONNX 模型
- 支持 INT8 量化加速

**客户端调用变更**：
- `transcribe.py` 新增 `--model` 和 `--fast` 参数
- `auto_transcribe.py` 新增 `--fast` 参数（单人快速模式）

---

## 优先级

- **P0**：Paraformer-large ONNX 加速（保持 diarization 能力，同时提速 2-4x）
- **P1**：SenseVoice-Small ONNX 单人快速模式（最速，但无 diarization）
- **P2**：单人模式自动路由（用户说单人讲课就自动走快速路径）

---

## 参考

- FunASR Issue #2706: SenseVoice + CAM++ compatibility
- FunASR Issue #2662: SenseVoice + CAM++ error  
- FunASR Issue #2121: Timestamp requirement for diarization
- funasr-onnx 包：https://pypi.org/project/funasr-onnx/

## 相关

- 当前版本：v1.5.1
- 路径：`skills/funasr-transcribe/`

### PR DESCRIPTION
**Pull request by cat-xierluo**
## Summary

- Adds and tunes the `paraformer-onnx` runtime path for `funasr-transcribe`, including ONNX compatibility export, ONNX VAD handling, text cleanup, punctuation restoration, and sentence-level timestamp reconstruction.
- Makes both single-speaker and multi-speaker ONNX paths use VAD-segmented ASR instead of whole-audio ONNX inference.
- Optimizes the single-speaker ONNX path with global punctuation restoration, while keeping multi-speaker punctuation segment-scoped for speaker alignment.
- Updates `funasr-transcribe` skill docs, CHANGELOG, TASKS/DECISIONS, API reference, README, and marketplace metadata.

## Root Cause

The initial single-speaker ONNX path sent long audio directly into the ONNX ASR model, while the multi-speaker path already benefited from VAD segmentation and post-processing. Whole-audio ONNX inference degraded both speed and transcription quality on long recordings.

## Validation

- `python3 -m py_compile skills/funasr-transcribe/scripts/server.py skills/funasr-transcribe/scripts/transcribe.py skills/funasr-transcribe/scripts/auto_transcribe.py`
- `python3 skills/funasr-transcribe/scripts/transcribe.py --help`
- `python3 skills/funasr-transcribe/scripts/auto_transcribe.py --help`
- `git diff --check origin/main...HEAD`
- 5-minute single-speaker lecture sample: optimized ONNX steady-state runtime about `11.751s`, raw text similarity about `0.9733`, normalized similarity about `0.9829` versus native Paraformer.
- 18-minute multi-speaker WeChat sample: optimized ONNX multi-speaker path about `291.332s`, versus native `paraformer + diarize` baseline about `551.59s`.
- 90-second multi-speaker ONNX smoke test produced sentence timestamps and speaker fields.

Closes #14

### CODE REVIEW

**Review by cat-xierluo — COMMENTED** [review]
## PR #16 审查：ONNX 转录优化

整体架构重构合理，文档同步非常完整，性能提升数据详实。以下是发现的问题：

---

### 阻塞性问题（建议合并前修复）

**1. `import torch` 和 `funasr.utils.export_utils` 在非 ONNX 路径上无条件导入**

`server.py` 在模块级别添加了 `import torch` 和 `import funasr.utils.export_utils`。即使只用原生 `paraformer`（不走 ONNX），这些也会被导入。`funasr.utils.export_utils` 是内部模块，路径可能随 FunASR 版本变化导致 import 失败。建议包裹在 `try/except` 中，或延迟到 `patch_funasr_onnx_export()` 首次调用时再导入。

**2. `patch_funasr_onnx_export()` 猴子补丁打到私有属性 `_onnx`**

`funasr_export_utils._onnx = compat_onnx_export` — `_onnx` 带下划线前缀是内部实现。如果 FunASR 更新重命名或移除该属性，补丁会**静默失效**，ONNX 导出走回有 bug 的原始路径，不会报错但产出错误结果。建议：
- 至少加 `hasattr` 检查并 `logging.warning`
- 或在 `ensure_compat_onnx_model_dir()` 中直接调用 `torch.onnx.export()` 绕过 FunASR 导出工具链

**3. `resolve_local_model_dir()` 有隐式网络 I/O**

函数名叫 "resolve" 但内部调用了 `snapshot_download()`，可能触发大模型下载。调用链 `get_onnx_compat_dir()` → `ensure_compat_onnx_model_dir()` 中没有明确提示。建议在下载前打印提示信息。

---

### 中等问题

**4. `call_onnx_vad_model()` 大量复制 funasr_onnx 内部逻辑**

约 50 行手动重现 VAD 完整推理管线，是为了修复 `feats_len` 标量/数组兼容问题。如果 funasr_onnx 更新内部 API，会静默产出错误结果。建议加版本检查注释标明补丁针对的版本，长期考虑提交上游 fix。

**5. `ensure_compat_onnx_model_dir()` 完整拷贝模型目录**

`shutil.copytree(source_dir, compat_dir, dirs_exist_ok=True)` 会把整个模型（可能数百 MB）复制到 `~/.cache/funasr-onnx-compat/`，无清理机制。建议只拷贝必要文件（模型权重 + config），或添加过期检查。

**6. `split_text_with_timestamps()` 按字符比例映射时间戳**

`start_char / total_chars * total_timestamps` 的比例映射假设每个字符语音时长相同，对中文+英文混排场景会产生时间偏移。建议在文档中标明这是近似映射。

**7. 版本号不一致（已有问题）**

`marketplace.json` 从 1.1.1 跳到 1.9.3，`README.md` 从 v1.2.0 跳到 v1.9.3，起始版本就不同。

---

### 小问题

**8.** `split_text_by_punctuation()` 未覆盖顿号 `、`、冒号 `：` 等中文标点，可能导致过长的"句子"。

**9.** `normalize_onnx_tokens()` 的 `@@` 子词续接逻辑针对 FunASR 特定 tokenizer，上游变更可能失效，建议加版本注释。

**10.** `transcribe_paraformer_onnx_segments()` 返回类型从 `dict` 改为 `tuple[dict, list]`，缺少类型标注。

---

### 优点

- 提取 `transcribe_paraformer_onnx_segments()` 作为单/多人共享逻辑，架构清晰
- 文档同步非常完整（SKILL.md、API reference、CHANGELOG、DECISIONS.md、TASKS.md）
- 决策记录详实（DEC-001 到 DEC-005）
- `--fast` 语义变更理由充分，有环境变量回退
- `join_text_fragments()` 替换 `' '.join()`，中文拼接更自然
- `auto_transcribe.py` 的 `start_server(api_url)` 修复是实际需要的

---

### 合并建议

建议 **squash merge**（同一天 3 轮迭代 1.9.1 → 1.9.3），合并前至少修复 #1 和 #2。

**Comment by cat-xierluo** [comment]
已在 `bcededc` 中处理这轮审查意见：

- 修复 #1：移除 `torch` / `funasr.utils.export_utils` 的模块级直接导入，改为 ONNX 兼容导出时按需加载。
- 修复 #2：对 `funasr.utils.export_utils._onnx` 增加存在性检查；FunASR 内部 API 变化时会直接抛清晰错误，不再静默回退。
- 处理 #3：模型缺失触发下载前会打印提示。
- 处理 #4/#9：补充 funasr-onnx VAD 兼容补丁和 raw token 处理的适用范围说明。
- 处理 #6：文档明确 ONNX 句子级时间戳是近似映射，不是逐字强对齐。
- 处理 #8/#10：扩展中文标点切分，并细化 `transcribe_paraformer_onnx_segments()` 返回类型标注。

关于 #5：暂未改成“只复制必要文件”。原因是 FunASR 导出链路依赖模型目录中的多类配置和资源，贸然裁剪容易破坏导出稳定性。本次改为增加 `compat_export_version` 缓存过期检查，并在日志/文档中说明缓存目录可删除后重建。

验证：

- `python3 -m py_compile skills/funasr-transcribe/scripts/server.py skills/funasr-transcribe/scripts/transcribe.py skills/funasr-transcribe/scripts/auto_transcribe.py`
- `python3 skills/funasr-transcribe/scripts/transcribe.py --help`
- `python3 skills/funasr-transcribe/scripts/auto_transcribe.py --help`
- `git diff --check`

**Review by cat-xierluo — COMMENTED** [review]
## PR #16 二次审查（v1.9.4 更新）

### 上次阻塞问题的修复状态

**#1 无条件 import — 已修复 ✅**
`torch` 和 `funasr.utils.export_utils` 已从模块级移除，改为 `load_onnx_export_dependencies()` 按需加载，非 ONNX 路径不再受影响。

**#2 猴子补丁静默失效 — 已修复 ✅**
新增 `hasattr(export_utils, "_onnx")` 检查，不存在时直接抛出 `RuntimeError`；同时通过 `LOGGER.warning` 记录补丁应用。新增 `ONNX_COMPAT_EXPORT_VERSION = 2` 缓存版本号，导出逻辑变化时旧缓存自动失效。

**#3 隐式网络 I/O — 已改善 ✅**
`resolve_local_model_dir()` 新增下载前打印提示。函数名仍叫 "resolve"，但下载前有提示，可接受。

### 中等问题修复状态

- **#4 VAD 兼容补丁** — docstring 标注 `funasr-onnx 1.3.x` ✅
- **#5 缓存磁盘占用** — 新增 `compat_export_version` + 清理提示 ✅
- **#6 时间戳近似映射** — SKILL.md、API reference、docstring 三处标注"近似映射" ✅

### 小问题修复状态

- **#8 标点覆盖** — 正则已加入 `、：` ✅
- **#9 tokenizer 注释** — docstring 标注 FunASR Paraformer ONNX ✅
- **#10 返回类型标注** — 添加 `-> tuple[dict, list[list[object]]]` ✅

### 结论

3 个阻塞问题全部修复，中等问题和小问题也已妥善处理。DECISIONS.md 新增 DEC-006 记录了修复决策。**LGTM，建议 squash merge。**
